### PR TITLE
[common] Simplified buffer

### PR DESCRIPTION
**Pull request by mimaric**
`buffer` was halfway between buffer and vector. It has been simplified now to be closer to a regular buffer.
The user explicitly specifies the size. If the size has to be changed old data is discarded (`clear_and_resize`). The buffer can also be move copied, move assigned and swapped. There is no concept of separate capacity and size anymore.

`host_buffer` has been removed. Content of `buffer` has been moved into `device_buffer` in order to simplify the file structure.

This PR also removes stream from `allocator`'s `deallocate()`.

Part of #414 and #318

### CODE REVIEW

File: common/base/include/claragenomics/utils/device_buffer.hpp [hunk 1]
```diff
@@ -10,20 +10,200 @@
...
+    {
+        assert(new_size >= 0);
+
+        if (new_size == size_)
+        {
+            return;
+        }
+
+        clear_and_free();
+        size_ = new_size;
```
Comment (ahehn-nv) [inline common/base/include/claragenomics/utils/device_buffer.hpp:177]: Optional: I know we are not really exception safe anyways, but assigning the size after the data, would at least guarantee that the buffer is of size 0 if the allocation fails.

File: common/base/include/claragenomics/utils/device_buffer.hpp [hunk 2]
```diff
@@ -10,20 +10,200 @@
...
+    cudaStream_t stream_;
+    Allocator allocator_;
+};
+
+} // namespace details
+
 template <typename T>
 #ifdef CGA_ENABLE_CACHING_ALLOCATOR
-using device_buffer = buffer<T, CachingDeviceAllocator<T, DevicePreallocatedAllocator>>;
+using device_buffer = details::buffer<T, CachingDeviceAllocator<T, DevicePreallocatedAllocator>>;
```
[the remark this answers is not in the document]
Comment (mimaric) [reply common/base/include/claragenomics/utils/device_buffer.hpp:204]: I'm don't think that's possible. `DefaultDeviceAllocator` is defined as `CachingDeviceAllocator<char, DevicePreallocatedAllocator>` (https://github.com/mimaric/ClaraGenomicsAnalysis/blob/311998dacffb717c188a88fab4b15b9f36652884/common/base/include/claragenomics/utils/allocator.hpp#L277) and `device_buffer` as `details::buffer<T, CachingDeviceAllocator<T, DevicePreallocatedAllocator>>` (https://github.com/mimaric/ClaraGenomicsAnalysis/blob/311998dacffb717c188a88fab4b15b9f36652884/common/base/include/claragenomics/utils/device_buffer.hpp#L204), so `char` and `T` do not play nicely.

Here's the error:
```
/home/.../common/base/include/claragenomics/utils/device_buffer.hpp(156): error: argument of type "uint32_t *" is incompatible with parameter of type "char *"
          detected during:
            instantiation of "void claraparabricks::genomeworks::details::buffer<T, Allocator>::clear_and_free() [with T=uint32_t, Allocator=claraparabricks::genomeworks::DefaultDeviceAllocator]"
/home/.../common/base/include/claragenomics/utils/cudasort.cuh(191): here
            instantiation of "void claraparabricks::genomeworks::cudautils::sort_by_two_keys(claraparabricks::genomeworks::device_buffer<MoreSignificantKeyT> &, claraparabricks::genomeworks::device_buffer<LessSignificantKeyT> &, claraparabricks::genomeworks::device_buffer<ValueT> &, MoreSignificantKeyT, LessSignificantKeyT, cudaStream_t) [with MoreSignificantKeyT=uint32_t, LessSignificantKeyT=uint32_t, ValueT=uint32_t]"
/home/.../common/base/tests/Test_UtilsCudasort.cu(46): here
            instantiation of "void claraparabricks::genomeworks::test_function(claraparabricks::genomeworks::device_buffer<MoreSignificantKeyT> &, claraparabricks::genomeworks::device_buffer<LessSignificantKeyT> &, claraparabricks::genomeworks::device_buffer<ValueT> &, MoreSignificantKeyT, LessSignificantKeyT) [with MoreSignificantKeyT=uint32_t, LessSignificantKeyT=uint32_t, ValueT=uint32_t]"
/home/.../common/base/tests/Test_UtilsCudasort.cu(100): here
            instantiation of "void claraparabricks::genomeworks::short_test_template_larger_more_significant_key<MoreSignificantKeyT,LessSignificantKeyT,ValueT>() [with MoreSignificantKeyT=uint32_t, LessSignificantKeyT=uint32_t, ValueT=uint32_t]"
/home/.../common/base/tests/Test_UtilsCudasort.cu(149): here
```

[the remark this answers is not in the document]
Comment (mimaric) [reply common/base/include/claragenomics/utils/device_buffer.hpp:204]: I guess we could make `DefaultDeviceAllocator` a template....

File: common/base/include/claragenomics/utils/device_buffer.hpp [hunk 3]
```diff
@@ -10,20 +10,200 @@
...
+    /// /brief This method returns the associated stream.
+    /// /return associated stream.
+    cudaStream_t get_stream() const { return stream_; }
+
+    /// /brief This method returns the allocator used to allocate memory for this buffer.
+    /// /return allocator.
+    allocator_type get_allocator() const { return allocator_; }
+
+    /// /brief This method releases the memory and resizes the buffer to 0.
+    void clear_and_free()
```
[the remark this answers is not in the document]
Comment (mimaric) [reply common/base/include/claragenomics/utils/device_buffer.hpp:152]: Both are fine for me. When I was renaming `resize()` into `clear_and_resize()` I also renamed `free()` into `clear_and_free()`, although I agree `clear()` implies free and `resize()` does not so it has to be stated explicitly.
I can rename it if you want.

**Comment by mimaric** [comment]
rerun tests

File: common/base/include/claragenomics/utils/device_buffer.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (ahehn-nv) [reply common/base/include/claragenomics/utils/device_buffer.hpp:152]: yes, please change it again.

File: common/base/include/claragenomics/utils/device_buffer.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (ahehn-nv) [reply common/base/include/claragenomics/utils/device_buffer.hpp:204]: Oh, damn. It's the second time I fall for that one. Fine to leave it as it is then...

File: common/base/include/claragenomics/utils/device_buffer.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (mimaric) [reply common/base/include/claragenomics/utils/device_buffer.hpp]: done

File: common/base/include/claragenomics/utils/device_preallocated_allocator.cuh [hunk 4]
```diff
@@ -133,7 +133,7 @@ private:
     }
 
     /// @brief finds a memory block of the given size
-    /// @param ptr
+    /// @param ptr on return pointer to allocated memory, nullptr is allocation was not successful
```
Comment (ahehn-nv) [inline common/base/include/claragenomics/utils/device_preallocated_allocator.cuh:136]: see above

File: common/base/include/claragenomics/utils/allocator.hpp [hunk 5]
```diff
@@ -105,7 +105,7 @@ class CudaMallocAllocator
     /// @brief asynchronously allocates a device array with enough space for n elements of value_type
     /// @param n number of elements to allocate the array for
     /// @param stream CUDA stream to be associated with this method
-    /// @return pointer to allocated array
+    /// @return pointer to allocated memory, nullptr is allocation was not successful
```
[the remark this answers is not in the document]
Comment (mimaric) [reply common/base/include/claragenomics/utils/allocator.hpp:108]: true, corrected

File: common/base/include/claragenomics/utils/allocator.hpp [hunk 6]
```diff
@@ -232,7 +230,7 @@ class CachingDeviceAllocator
...
```
[the remark this answers is not in the document]
Comment (mimaric) [reply common/base/include/claragenomics/utils/allocator.hpp:233]: corrected

File: common/base/include/claragenomics/utils/device_preallocated_allocator.cuh [hunk 7]
```diff
@@ -76,7 +76,7 @@ public:
 
     /// @brief allocates memory (assigns part of the buffer)
     /// Memory allocation is aligned by 256 bytes
-    /// @param ptr
+    /// @param ptr on return pointer to allocated memory, nullptr is allocation was not successful
```
[the remark this answers is not in the document]
Comment (mimaric) [reply common/base/include/claragenomics/utils/device_preallocated_allocator.cuh]: done

File: common/base/include/claragenomics/utils/device_preallocated_allocator.cuh (hunk 4 above)
Comment (mimaric) [reply common/base/include/claragenomics/utils/device_preallocated_allocator.cuh]: done

File: common/base/include/claragenomics/utils/device_preallocated_allocator.cuh [hunk 8]
```diff
@@ -74,12 +74,12 @@ public:
...
     // ^^^^ buffer_'s destructor deallocates device memory
 
-    /// @brief allocates memory (assigns part of the buffer)
+    /// \brief allocates memory (assigns part of the buffer)
     /// Memory allocation is aligned by 256 bytes
-    /// @param ptr
-    /// @param bytes_needed
-    /// @param associated_stream on deallocation this block will be free only once all work in this stream has finished
-    /// @return error status
+    /// \param ptr on return pointer to allocated memory, nullptr is allocation was not successful
```
Comment (ahehn-nv) [inline common/base/include/claragenomics/utils/device_preallocated_allocator.cuh:79]: typo is->if

File: common/base/include/claragenomics/utils/device_preallocated_allocator.cuh [hunk 9]
```diff
@@ -132,15 +132,17 @@ private:
         return ret_val;
     }
 
-    /// @brief finds a memory block of the given size
-    /// @param ptr
-    /// @param bytes_needed
-    /// @param associated_stream On deallocation this block will be free only once all work in this stream has finished
-    /// @return error status
+    /// \brief finds a memory block of the given size
+    /// \param ptr on return pointer to allocated memory, nullptr is allocation was not successful
```
Comment (ahehn-nv) [inline common/base/include/claragenomics/utils/device_preallocated_allocator.cuh:136]: typo is -> if

File: common/base/include/claragenomics/utils/allocator.hpp [hunk 10]
```diff
@@ -37,75 +37,76 @@ class CudaMallocAllocator
...
 
-    /// @brief asynchronously allocates a device array with enough space for n elements of value_type
-    /// @param n number of elements to allocate the array for
-    /// @param stream CUDA stream to be associated with this method
-    /// @return pointer to allocated array
+    /// \brief asynchronously allocates a device array with enough space for n elements of value_type
+    /// \param n number of elements to allocate the array for
+    /// \param stream CUDA stream to be associated with this method
+    /// \return pointer to allocated memory
+    /// \throw device_memory_allocation_exception is allocation was not successful
```
Comment (ahehn-nv) [inline common/base/include/claragenomics/utils/allocator.hpp:109]: typo is -> if

File: common/base/include/claragenomics/utils/allocator.hpp [hunk 11]
```diff
@@ -151,88 +150,89 @@ class CachingDeviceAllocator
...
```
Comment (ahehn-nv) [inline common/base/include/claragenomics/utils/allocator.hpp:235]: typo: is -> if

Comment (mimaric) [reply common/base/include/claragenomics/utils/allocator.hpp]: done

File: common/base/include/claragenomics/utils/allocator.hpp (hunk 10 above)
Comment (mimaric) [reply common/base/include/claragenomics/utils/allocator.hpp]: done

File: common/base/include/claragenomics/utils/device_preallocated_allocator.cuh (hunk 9 above)
Comment (mimaric) [reply common/base/include/claragenomics/utils/device_preallocated_allocator.cuh]: done

File: common/base/include/claragenomics/utils/device_preallocated_allocator.cuh (hunk 8 above)
Comment (mimaric) [reply common/base/include/claragenomics/utils/device_preallocated_allocator.cuh]: done

File: common/base/include/claragenomics/utils/device_preallocated_allocator.cuh [hunk 12]
```diff
@@ -74,12 +74,12 @@ public:
...
+    /// \brief allocates memory (assigns part of the buffer)
     /// Memory allocation is aligned by 256 bytes
-    /// @param ptr
-    /// @param bytes_needed
-    /// @param associated_stream on deallocation this block will be free only once all work in this stream has finished
-    /// @return error status
+    /// \param ptr on return pointer to allocated memory, nullptr if allocation was not successful
+    /// \param bytes_needed
+    /// \param associated_stream on deallocation this block will be free only once all work in this stream has finished
+    /// \return cudaSuccess is allocation was successful, cudaErrorMemoryAllocation otherwise
```
Comment (ahehn-nv) [inline common/base/include/claragenomics/utils/device_preallocated_allocator.cuh:82]: the typo strikes again...

File: common/base/include/claragenomics/utils/device_preallocated_allocator.cuh [hunk 13]
```diff
@@ -132,15 +132,17 @@ private:
...
-    /// @brief finds a memory block of the given size
-    /// @param ptr
-    /// @param bytes_needed
-    /// @param associated_stream On deallocation this block will be free only once all work in this stream has finished
-    /// @return error status
+    /// \brief finds a memory block of the given size
+    /// \param ptr on return pointer to allocated memory, nullptr if allocation was not successful
+    /// \param bytes_needed
+    /// \param associated_stream On deallocation this block will be free only once all work in this stream has finished
+    /// \return cudaSuccess is allocation was successful, cudaErrorMemoryAllocation otherwise
```
Comment (ahehn-nv) [inline common/base/include/claragenomics/utils/device_preallocated_allocator.cuh:139]: ... and again. :D

File: common/base/include/claragenomics/utils/device_preallocated_allocator.cuh (hunk 12 above)
Comment (mimaric) [reply common/base/include/claragenomics/utils/device_preallocated_allocator.cuh]: done

File: common/base/include/claragenomics/utils/device_preallocated_allocator.cuh (hunk 13 above)
Comment (mimaric) [reply common/base/include/claragenomics/utils/device_preallocated_allocator.cuh]: done

**Pull request merged**